gs_quant/api/fred/fredapi.py

```diff
@@ -0,0 +1,332 @@
+"""
+Copyright 2019 Goldman Sachs.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+
+//Portions copyright Maverick Lin. Licensed under Apache 2.0 license
+"""
+
+import os
+import sys
+import json
+import requests
+from urllib.parse import urlencode
+from requests.utils import requote_uri
+import datetime as dt
+from requests.exceptions import HTTPError
+from typing import Iterable, List, Optional, Tuple, Union
+
+import pandas as pd
+
+"""
+FRED API that provides useful functions to query the Fred dataset. You need to specify a valid
+API key in one of 3 ways: pass the string via api_key, or set api_key_file to a file with the api key in the
+first line, or set the environment variable 'FRED_API_KEY' to the value of your api key. You can sign up for a
+free api key on the Fred website at http://research.stlouisfed.org/fred2/
+"""
+
+class FredDataApi(object):
+    earliest_realtime_start = '1776-07-04'
+    latest_realtime_end = '9999-12-31'
+    root_url = 'https://api.stlouisfed.org/fred'
+
+    def __init__(self, api_key=None, api_key_file=None):
+        self.api_key = api_key
```

Comment:
I don’t think we need the API key file – lets just pass it as a string; it will simplify the init significantly